File: backend/src/ase/api/routers/teams.py
"""Team management; the application service enforces account and membership authority."""

from uuid import UUID

from fastapi import APIRouter, Response

from ase.api.deps import ContainerDep, ContextDep, CurrentUser, SessionDep
from ase.api.schemas_ai_usage import TeamAiUsageOut
from ase.api.team_schemas import (
    MemberIn,
    MemberOut,
    MemberRoleIn,
    MembershipOut,
    TeamDetailOut,
    TeamIn,
    TeamOut,
    TeamsOut,
    TeamUpdateIn,
)
from ase.application.teams.service import DESCRIPTION_UNSET

router = APIRouter(prefix="/teams", tags=["teams"])


@router.get("")
async def list_teams(user: CurrentUser, session: SessionDep, container: ContainerDep) -> TeamsOut:
    items = await container.teams(session).list_teams(user)
    return TeamsOut(items=[TeamOut.model_validate(team) for team in items])


@router.post("", status_code=201)
async def create_team(
    body: TeamIn,
    user: CurrentUser,
    session: SessionDep,
    container: ContainerDep,
    context: ContextDep,
) -> TeamOut:
    team = await container.teams(session).create(
        user, body.name, context, description=body.description
    )
    return TeamOut.model_validate(team)


@router.get("/{team_id}")
async def get_team(
    team_id: UUID, user: CurrentUser, session: SessionDep, container: ContainerDep
) -> TeamDetailOut:
    team, members = await container.teams(session).roster(user, team_id)
    return TeamDetailOut(
        team=TeamOut.model_validate(team),
        members=[MemberOut.model_validate(member) for member in members],
    )


@router.get("/{team_id}/ai-usage")
async def team_ai_usage(
    team_id: UUID,
    user: CurrentUser,
    session: SessionDep,
    container: ContainerDep,
    response: Response,
) -> TeamAiUsageOut:
    # Members see their own attributed usage, Managers see aggregates and member
    # totals, administrators see any team. Non-members receive 404.
    response.headers["Cache-Control"] = "no-store"
    usage = await container.ai_usage_views(session).team(user, team_id)
    return TeamAiUsageOut.from_usage(usage)


@router.patch("/{team_id}")
async def update_team(
    team_id: UUID,
    body: TeamUpdateIn,
    user: CurrentUser,
    session: SessionDep,
    container: ContainerDep,
    context: ContextDep,
) -> TeamOut:
    team = await container.teams(session).update(
        user,
        team_id,
        name=body.name,
        is_active=body.is_active,
        description=(
            body.description if "description" in body.model_fields_set else DESCRIPTION_UNSET
        ),
        context=context,
    )
    return TeamOut.model_validate(team)


@router.put("/{team_id}/members")
async def set_member(
    team_id: UUID,
    body: MemberIn,
    user: CurrentUser,
    session: SessionDep,
    container: ContainerDep,
    context: ContextDep,
) -> MembershipOut:
    member = await container.teams(session).set_member(
        user, team_id, email=str(body.email), role=body.role, context=context
    )
    return MembershipOut.model_validate(member)


@router.patch("/{team_id}/members/{user_id}")
async def change_member_role(
    team_id: UUID,
    user_id: UUID,
    body: MemberRoleIn,
    user: CurrentUser,
    session: SessionDep,
    container: ContainerDep,
    context: ContextDep,
) -> MembershipOut:
    member = await container.teams(session).change_role(user, team_id, user_id, body.role, context)
    return MembershipOut.model_validate(member)


@router.post("/{team_id}/leave", status_code=204, response_class=Response)
async def leave_team(
    team_id: UUID,
    user: CurrentUser,
    session: SessionDep,
    container: ContainerDep,
    context: ContextDep,
) -> Response:
    await container.teams(session).leave(user, team_id, context)
    return Response(status_code=204)


@router.delete("/{team_id}/members/{user_id}", status_code=204, response_class=Response)
async def remove_member(
    team_id: UUID,
    user_id: UUID,
    user: CurrentUser,
    session: SessionDep,
    container: ContainerDep,
    context: ContextDep,
) -> Response:
    await container.teams(session).remove_member(user, team_id, user_id, context)
    return Response(status_code=204)

File: frontend/src/lib/api/teams.ts
/** Team contracts come from OpenAPI; responses are validated before rendering. */
import { z } from 'zod';

import { apiCall, apiSend } from './client';
import type { components } from './types.gen';

export type Team = components['schemas']['TeamOut'];
export type TeamDetail = components['schemas']['TeamDetailOut'];
export type TeamMember = components['schemas']['MemberOut'];
export type MemberInput = components['schemas']['MemberIn'];
export type MemberRole = components['schemas']['MemberRoleIn']['role'];
export type TeamUpdate = components['schemas']['TeamUpdateIn'];

const teamSchema = z.object({
  id: z.uuid(),
  name: z.string(),
  is_active: z.boolean(),
  created_by: z.uuid(),
  created_at: z.string(),
  updated_at: z.string(),
  // Older deployments may omit the optional description while migrations are
  // being rolled out. Normalise that response to the contract's null value.
  description: z.string().nullable().default(null),
}) satisfies z.ZodType<Team>;

const teamsSchema = z.object({ items: z.array(teamSchema) }) satisfies z.ZodType<
  components['schemas']['TeamsOut']
>;
const detailSchema = z.object({
  team: teamSchema,
  members: z.array(
    z.object({
      user_id: z.uuid(),
      display_name: z.string(),
      // Rosters carry the directory handle only, never a login email address.
      username: z.string().nullable(),
      account_role: z.enum(['user', 'manager', 'admin']),
      is_active: z.boolean(),
      role: z.enum(['member', 'manager']),
      joined_at: z.string(),
    }),
  ),
}) satisfies z.ZodType<TeamDetail>;
const membershipSchema = z.object({
  team_id: z.uuid(),
  user_id: z.uuid(),
  role: z.enum(['member', 'manager']),
  joined_at: z.string(),
}) satisfies z.ZodType<components['schemas']['MembershipOut']>;

export async function listTeams(): Promise<Team[]> {
  return (await apiCall('/api/teams', { schema: teamsSchema })).items;
}
export function getTeam(id: string): Promise<TeamDetail> {
  return apiCall(`/api/teams/${encodeURIComponent(id)}`, { schema: detailSchema });
}
export function createTeam(name: string, description?: string): Promise<Team> {
  const body: components['schemas']['TeamIn'] = description ? { name, description } : { name };
  return apiCall('/api/teams', { method: 'POST', body, schema: teamSchema });
}
export function updateTeam(id: string, body: TeamUpdate): Promise<Team> {
  return apiCall(`/api/teams/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body,
    schema: teamSchema,
  });
}
/** Administrator direct add. Team managers invite people instead. */
export function setMember(id: string, body: MemberInput) {
  return apiCall(`/api/teams/${encodeURIComponent(id)}/members`, {
    method: 'PUT',
    body,
    schema: membershipSchema,
  });
}
/** Promote or demote an existing member by account id. */
export function changeMemberRole(id: string, userId: string, role: MemberRole) {
  const body: components['schemas']['MemberRoleIn'] = { role };
  return apiCall(`/api/teams/${encodeURIComponent(id)}/members/${encodeURIComponent(userId)}`, {
    method: 'PATCH',
    body,
    schema: membershipSchema,
  });
}
/** Leave a team. The server refuses if it would remove the last active manager. */
export function leaveTeam(id: string): Promise<void> {
  return apiSend(`/api/teams/${encodeURIComponent(id)}/leave`, { method: 'POST' });
}
export function removeMember(id: string, userId: string): Promise<void> {
  return apiSend(`/api/teams/${encodeURIComponent(id)}/members/${encodeURIComponent(userId)}`, {
    method: 'DELETE',
  });
}
